### PR TITLE
perf(secrets): probe a payload for each pattern's required literals before scanning it

### PR DESCRIPTION
**Pull request by alexander-turner**
**Claimed area:** `agent_sanitizer.secrets` redaction hot path (`engine.py`, `invisible.py`, new `prefilter.py`).

Follows #350, which has merged — base is now `main` and this diff is only the rewrite.

## Summary

#350 stopped `scan_line` from running on every line. What was left over was the same problem one level up: the engine still ran a per-line union of all 57 denylists, a PEM `sub` over the whole payload, a field-value `sub` over the whole payload, and the cross-line union's `finditer` over the payload's newline-free collapse — none of which any given input can usually match.

Every one of those is now gated on a **necessary condition derived from the pattern's own parsed AST**: a literal string that every match of the pattern must contain, searched with `str.find`. The new `prefilter.LiteralProbe` also **indexes** each literal back to the patterns that named it, so a candidate line is confirmed against only those — a line carrying the word `key` tests the keyword detector, not the other 56 denylists.

Detection is unchanged by construction and by test: a literal the AST walk fails to derive means more text gets scanned, never less, and `test_redaction_is_byte_identical_with_and_without_the_probes` asserts the output and reported types match an engine with every probe opened.

Measured (same corpora both sides, #350's head vs. this branch):

| payload | before | after |
| --- | --- | --- |
| 12.9MB synthetic log, 240k lines | 22.9s | **1.2s** |
| 7.2MB of this repo's own source, 191k lines | 8.8s | **5.9s** |
| 1MiB single line of `key ` | 13s | **0.94s** |

The source-shaped payload stays the slower of the two on purpose: `key`/`secret`/`token` genuinely appear on ~20% of source lines, so those lines are real candidates and have to be scanned. Both shapes now scale linearly (1/2/4 MiB single line → 0.95/1.86/3.74s; 0.5/1/2 MiB of keyword-per-line → 1.37/2.77/5.64s).

**Review focus:** `python/agent_sanitizer/secrets/prefilter.py` — `_required()` (the AST walk; every opcode it treats as contributing nothing is a soundness claim, and a negative lookaround read as *required* would be the bug) and `_CASE_EQUIVALENT_FIXUPS` (five code points `re.IGNORECASE` treats as equal to a literal character that `str.lower` does not map onto it — the probe folds its haystack, so missing one is a missed secret; two brute-force tests over all of Unicode derive the table rather than trusting it, one per code point and one for context-freeness). The part I'm least sure of is the `weak` path: a pattern whose only required literal is one character is run as a regex over the whole text, and because `finditer` is non-overlapping a match spanning a newline must claim *every* line it touches, not just the one it starts on.

## Changes

- **New `prefilter.py`.** `required_literals()` walks a pattern's `re._parser` AST (with the `sre_parse` spelling for the 3.10 floor) and returns strings any match must contain. `LiteralProbe` splits a pattern set into a literal→patterns index plus the patterns too weak to index, and answers two questions: `may_match(text)` for the whole-payload passes, and `candidates(text, check)` mapping each candidate line index to the patterns that could match it.
- **Linear attribution.** `_patterns_by_line` carries the next newline's position across hits instead of re-searching per hit, and each literal reports one hit per **line** rather than one per occurrence (every occurrence of a fixed literal starting on the same line spans the same lines, so this is exact). Identical `(line, patterns)` claims merge once. Without these, one long line with a hit per word rescanned its whole remaining text per hit.
- **The probe honours the compute budget.** `compute_budget_seconds` is threaded into `candidates()` and checked between its units of work (per literal, per weak pattern, before attribution), each of which is linear in the payload. Previously the deadline was checked only *before* the stage, so nothing could interrupt it.
- **`engine.py`:** `_line_probe()` / `_eligible_probe()` (plugin-derived, cached and cleared with the existing prefilters) and module-level `_PEM_PROBE` / `_FIELD_VALUE_PROBE` derived from the very patterns they gate. `_redact_line` takes a `confirm` tuple; `_redact_lines` skips a non-candidate line without calling it. `_full_prefilter` stays as the unnarrowed fallback and as `confirm`'s default, because an over-wide `confirm` costs time while a too-narrow one silently stops detecting.
- **`invisible.py`:** `DeletionOffsets` replaces the per-character offset list (one entry per *deletion*, binary-searched) and `_invisible_class` range-collapses the 438-code-point charset, which also shortens the env-value run pattern. The old code built a 6M-element Python list and scanned with a 438-alternative character class.
- `_degroup` moves to `prefilter.degroup` (single copy; the engine imports it), and the fail-loud "plugin supplied no denylist" check moves to `prefilter.denylist_patterns`, shared by the prefilters and the probes.
- One existing test's compute budget raised from 1ms to 50ms: the whole-payload probe now runs before the per-line loop, so a 1ms budget expired during it and the test stopped asserting on the stage it names.

## Tests / verification

- New `tests/secrets/test_literal_probe.py` (82 tests): per-opcode AST-walk cases; the required-literal property checked against every live denylist over the detector sample corpus; two Unicode brute-forces deriving the case-fold table (per-code-point equivalence, and context-freeness); `fold` length-preservation over all code points; the literal→pattern index and its bounds; the weak-pattern newline-span claim; a per-probe non-vacuity guard; the attribution's search-count bound; the deadline wiring; fail-loud on an empty probe and on a denylist-less plugin; and the byte-identical differential over six payloads including this engine's own source.
- Non-vacuity checked by injecting bugs and confirming red: dropping the weak patterns (5 differential cases fail), removing the case fixups (2 fold tests fail), claiming only a weak match's starting line (1 test fails), reverting the attribution walk (24002 `str.find` calls vs. 3, with identical output), and resuming the per-line skip past an occurrence's END rather than its start (a multi-line literal drops the line its real match sits on).
- `uv run --extra dev pytest tests/secrets` — 2201 passed, 5 skipped. `ruff check` / `ruff format --check` clean.

## Decisions made

- **`LITERAL_MIN_LEN = 2`, not 3.** At 3 the two-character prefixes (AWS's `KIA`/`SIA`, GitLab's `gl`) fall back to whole-text regexes: the source payload gets faster but the log payload gets ~75% slower. 2 is the better of the two; 4+ is far worse on both.
- **U+0130 is folded to `i` rather than declining.** `str.lower` expands it to two characters, which would desynchronize every offset from its line, so `fold` could have returned `None` and made the caller scan everything. That is correct but hands any Turkish-text payload a 10x cliff — enough to trip the daemon's compute budget and have the request refused. Mapping it one-to-one is equally sound (`re.IGNORECASE` does match `i` against it) and has no cliff. The length guard stays as the invariant, with a test proving no code point now reaches it.
- **Both sigmas fold onto U+03C3 rather than restricting the index to ASCII.** `str.lower` maps a word-final capital sigma differently from any other, the one lowering in Python that reads a character's neighbours — so a literal folded in isolation could miss its own occurrence in a folded haystack. Excluding non-ASCII literals from the index would also close it, but `contraseña` *is* a required literal of six live keyword denylists, so that would push the two most expensive detectors onto whole-payload regexes. Collapsing the sigmas keeps the fold 1:1 and length-preserving, and all three sigmas are `re.IGNORECASE`-equivalent so nothing is lost.
- **The cross-line union is gated but not windowed.** `may_match` skips its whole sweep for a payload with no structural literal, which is what makes the log case fast, but a payload that has one anywhere still pays the full sweep. Windowing around each literal hit needs a max-match-length bound computed from the AST; left as follow-up.

## Checklist

- [x] Commits follow Conventional Commits; each subject reads as a user-facing release note.
- [x] `pytest`, `ruff check`, and `ruff format` pass locally.
- [x] No detector was added or changed — the probes are necessary conditions with their own soundness tests, and the differential test asserts zero change in findings over a corpus of legitimate source.
- [x] No real secrets/tokens in the diff, tests, or description.
- [x] `CHANGELOG.md` and `package.json` version are left untouched.

### CODE REVIEW

File: python/agent_sanitizer/secrets/prefilter.py [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+    Walks the newlines once in span-start order rather than calling ``str.count``
+    per span (quadratic) or materializing a line-start table as long as the file
+    even when two lines matched.
+    """
+    claims: dict[int, set[re.Pattern[str]]] = {}
+    line = 0
+    cursor = 0
+    for start, end, patterns in sorted(hits, key=lambda hit: hit[0]):
+        while True:
+            nxt = text.find("\n", cursor)
```
[the remark this answers is not in the document]
Comment (alexander-turner) [reply python/agent_sanitizer/secrets/prefilter.py]: Confirmed and fixed in 6f93055. Three changes:

- `_patterns_by_line` now carries the next newline's position (and the fact that there is none) across hits, so the walk is monotone — 24002 `str.find` calls became 3 on a 3-line × 4000-hit payload, with byte-identical output.
- Each literal now reports one hit per **line** rather than one per occurrence: after a hit, the scan resumes past the following newline. Every occurrence of a fixed literal starting on the same line spans the same lines, so this is exact.
- `compute_budget_seconds` is threaded into `LiteralProbe.candidates` and checked between its units of work (per literal, per weak pattern, before attribution), so the budget can now interrupt the stage.

1 MiB single line of `key `: 13s → 0.94s. Scaling is now linear in both directions that were suspect (1/2/4 MiB single line → 0.95/1.86/3.74s; 0.5/1/2 MiB of keyword-per-line → 1.37/2.77/5.64s). Pinned by `test_line_attribution_searches_the_newlines_once_not_once_per_hit` (bounded on search count, not wall clock), `test_a_literal_is_reported_once_per_line_not_once_per_occurrence`, and `test_the_engine_hands_the_probe_the_requests_own_deadline`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: python/agent_sanitizer/secrets/prefilter.py [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+    distinct leading character class lets ``re`` skip ahead, which an alternation
+    defeats.
+    """
+
+    def __init__(self, patterns: Iterable[re.Pattern[str]]) -> None:
+        by_literal: dict[str, list[re.Pattern[str]]] = {}
+        weak: list[re.Pattern[str]] = []
+        for pattern in patterns:
+            found = required_literals(pattern.pattern, pattern.flags)
+            if found is None or min(len(lit) for lit in found) < LITERAL_MIN_LEN:
```
[the remark this answers is not in the document]
Comment (alexander-turner) [reply python/agent_sanitizer/secrets/prefilter.py:244]: The hazard is real — U+03A3 is the only counterexample over all 0x110000 code points, confirmed by brute force — but the suggested fix does not hold: the premise "every indexed literal is ASCII" is false today. Six of the keyword denylists derive `contraseña` as a required literal (it is an arm of the `password|passwd|...|contraseña|contrasena` alternation), so `not all(lit.isascii())` would drop all six keyword patterns to `weak` — meaning the two most expensive detectors sweep the whole payload as regexes on every request, which is the cost this PR exists to remove.

Fixed at the fold instead, in 6f93055: `_CASE_EQUIVALENT_FIXUPS` now maps both U+03A3 and U+03C2 onto U+03C3 ahead of `.lower()`, which makes the fold context-free while staying 1:1 and length-preserving. All three sigmas are `re.IGNORECASE`-equivalent, so collapsing them loses no equivalence.

Your methodology point is the load-bearing half, and it is now pinned rather than argued: `test_the_fold_is_context_free_over_all_of_unicode` asserts `fold(before + c + after) == before + fold(c) + after` for every code point in two contexts. It goes red on the pre-fix table, which is what makes the per-code-point test a complete derivation regardless of what vocabulary detect-secrets grows.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: python/agent_sanitizer/secrets/engine.py [hunk 3]
```diff
@@ -1526,21 +1552,33 @@ def _redact_lines(
...
     cache: dict[str, tuple[str, list[str]]] = {}
     redacted_lines: list[str] = []
-    for line in lines:
+    for index, line in enumerate(lines):
         deadline.check("per-line scan")
+        if candidates is not None and index not in candidates:
+            redacted_lines.append(line)
+            continue
+        confirm = None if candidates is None else candidates[index]
         cached = cache.get(line)
```
[the remark this answers is not in the document]
Comment (alexander-turner) [reply python/agent_sanitizer/secrets/engine.py:1582]: Applied in 6f93055, with the break condition named: the comment states the two-line invariant and adds that a literal spanning a newline would violate it and must widen the key.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/secrets/test_literal_probe.py [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+
+
+# ─── the case fold ───────────────────────────────────────────────────────────
+
+
+def _probe_literal_chars() -> set[str]:
+    with E.configure_plugins():
+        probes = [E._line_probe(), E._eligible_probe()]
+    probes += [E._PEM_PROBE, E._FIELD_VALUE_PROBE]
+    return {char for probe in probes for lit in probe.by_literal for char in lit}
```
[the remark this answers is not in the document]
Comment (alexander-turner) [reply tests/secrets/test_literal_probe.py]: Applied in 6f93055 as `test_every_engine_probe_indexes_at_least_one_literal`, with the four probes hoisted into a shared `_engine_probes()` helper that `_probe_literal_chars` now reads, so the union and the per-probe assertion cannot drift apart.

---
_Generated by [Claude Code](https://claude.ai/code)_
